drivers: i3c: npcx: introduce I3C driver

alvsun opened the pull request:
This implements basic driver to utilize the I3C IP block on NPCX.

1. I3C mode: Main controller mode only.
2. Transfer: Support SDR only.
3. IBI: Support Hot-Join, IBI(MDB). Controller request is not supported.
4. Support 3 I3C modules: I3C1(3.3V), I3C2(1.8V, espi mode), (I3C3 1.8V or 3.3V)

alvsun commented:
This PR depends on #70932.
After #70932 is merged into ToT, we will begin the review.

dcpleung commented:
Could you split this into separate commits? I can see the I3C driver change, clock control change, and SoC/board change.

alvsun committed:
drivers: clock_control: npcx: add MCLKD as i3c source clock

1. The only valid values of MCLKD clock frequency
   are between 40Mhz to 50Mhz.
2. If DMA is used, the APB4_CLK clock frequency must
   be equal to or higher than 20Mhz.

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

alvsun committed:
dts: arm: nuvoton: add I3C device nodes

Add I3C device nodes.

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

alvsun committed:
boards: nuvoton: npcx4m8f_evb: add i3c pinctrl

Add I3C pinmux configuration.

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

alvsun commented:
> Could you split this into separate commits? I can see the I3C driver change, clock control change, and SoC/board change.

Done

XenuIsWatching commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2047 @@
+		return -EINVAL;
+	}
+
+	if (config->common.dev_list.num_i3c == 0) {
+		/*
+		 * No i3c devices in dev tree. Just return so
+		 * we don't get errors doing cmds when there
+		 * are no devices listening/responding.
+		 */
+		return 0;
+	}
```
I'm not sure what the point of this is? I think someone want to see the nack on the i3c broadcast header if there are not i3c devices currently known in it's 'list.

I think you're doing this because of the `i3c_bus_init` in the driver init function so you will a return of `0` rather than an error. Would it make more sense to have it "not call" the `i3c_bus_init` if there is nothing in it's i3c device list instead?

alvsun replied:
Done.
I have moved the target device checking before `i3c_bus_init` in `npcx_i3c_init`.

XenuIsWatching commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2047 @@
+
+struct npcx_i3c_data {
+	struct i3c_driver_data common;       /* Common i3c driver data */
+	struct k_sem lock_sem;               /* Mutex of i3c controller */
```
shouldn't this be a 'mutex' to prevent a priority inversion

alvsun replied:
Done, thanks

XenuIsWatching reviewed (changes requested):
Looks Great! only a couple of questions I have here

fabiobaltieri commented on drivers/i3c/i3c_common.c:
this file should be in its own commit, usually a good idea to avoid mixing driver and subsystem changes unless the changes are related (like it would break the build to not submit them together)

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(npcx_i3c, CONFIG_I3C_LOG_LEVEL);
+
+#define NPCX_I3C_CHK_TIMEOUT 10000    /* Unit: us, timeout for checking register status */
+#define I3C_SCL_PP_FREQ_MAX  12500000 /* Unit: Mhz*/
+#define I3C_SCL_OD_FREQ_MAX  4170000  /* Unit: Mhz*/
```
use the unit in the define naming, drop the comment, `I3C_SCL_OD_FREQ_MAX_MHZ` etc

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+#define BAMATCH_DIV     0x4 /* BAMATCH = APB4_CLK divided by four */
+
+/* Default maximum time we allow for an I3C transfer (unit:ms) */
+#define I3C_TRANS_TIMEOUT K_MSEC(100)
```
I would spec this in MS (`I3C_TRANS_TIMEOUT_MS`) and use `K_MSEC` in the usage, just cor coherency with how other time units in the file are declared

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	if (k_sem_take(&data->sync_sem, I3C_TRANS_TIMEOUT) == 0) {
+		return 0;
+	}
+
+	return -ETIMEDOUT;
```
could you just `return k_sem_take(...)`?

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+static bool npcx_i3c_has_error(struct i3c_reg *inst)
+{
+	if (IS_BIT_SET(inst->MSTATUS, NPCX_I3C_MSTATUS_ERRWARN)) {
+
```
nit: drop the empty line here

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+{
+	if (IS_BIT_SET(inst->MSTATUS, NPCX_I3C_MSTATUS_ERRWARN)) {
+
+		LOG_DBG("ERROR: MSTATUS 0x%08x MERRWARN 0x%08x", inst->MSTATUS, inst->MERRWARN);
```
should this be a `LOG_ERR` or `LOG_WRN`?

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+	ret = npcx_i3c_send_request(inst, mctrl);
+	if (ret != 0) {
+		LOG_ERR("Request start error, %d", ret);
+	}
```
return `err`? (and `return 0` at the end)

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+static inline int npcx_i3c_request_emit_stop(struct i3c_reg *inst)
+{
+	uint32_t val = 0;
+	int ret = 0;
```
no need to initialize `ret`, same in the other functions below, there's a few of these

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+	LOG_DBG("MIBIRULES 0x%08x", ibi_rules);
+}
+
+int npcx_i3c_ibi_enable(const struct device *dev, struct i3c_device_desc *target)
```
`static`

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+	return ret;
+}
+
+int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc *target)
```
`static`

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+	/* MCLKD = MCLK / I3C_DIV(1 or 2)
+	 * MCLKD must between 40 mhz to 50 mhz.
+	 */
+	if (mclkd == MHZ(45)) {
```
add a define for that 45MHZ (fancy, never realized we had that helper until now)

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+	static struct npcx_i3c_data npcx_i3c_data_##id = {                                         \
+		.common.ctrl_config.scl.i3c = DT_INST_PROP_OR(id, i3c_scl_hz, 0),                  \
+		.common.ctrl_config.scl.i2c = DT_INST_PROP_OR(id, i2c_scl_hz, 0),                  \
+		.clocks.i3c_od_scl_hz = DT_INST_PROP_OR(id, i3c_od_scl_hz, 0),                     \
```
should this be in `struct npcx_i3c_config`? does not looks like you are changing it in runtime

if that's the case then you are left with the other two, I would consider saving those in the config structure as well and initializing the data copy at init time from the code, would save the flash imagine from having an entire copy of the data structure just to intiialize two fields at boot time and the rest with zero

XenuIsWatching commented:
With https://github.com/zephyrproject-rtos/zephyr/pull/70773 now merged... you'll need to add the compact macro to https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L88 and https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L99 in a separate commit

alvsun committed:
drivers: i3c: update i3c_dev_list_daa_addr_helper()

During DAA, the responding device might not be in the device list.
This CL adds target device descriptor's pointer checking to prevent
getting unexpected results.

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

alvsun committed:
drivers: i3c: npcx: introduce NPCX I3C driver

This implements basic driver to utilize the I3C IP block
on NPCX.

1. I3C mode: Main controller mode only.
2. Transfer: Support SDR only.
3. IBI: Support Hot-Join, IBI(MDB).
   Controller request is not supported.
4. Support 3 I3C modules:
   I3C1(3.3V), I3C2(1.8V, espi mode), (I3C3 1.8V or 3.3V)

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

alvsun committed:
drivers: i3c: i3c_shell: add nuvoton_npcx_i3c compat macro to i3c_shell.

As titile.

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

alvsun commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+
+	ret = npcx_i3c_send_request(inst, val);
+	if (ret != 0) {
+		LOG_ERR("Request auto ibi error, %d", ret);
```
Done

alvsun commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+{
+	uint32_t mask = (BIT(NPCX_I3C_MSTATUS_TGTSTART) | BIT(NPCX_I3C_MSTATUS_MCTRLDONE) |
+			 BIT(NPCX_I3C_MSTATUS_COMPLETE) | BIT(NPCX_I3C_MSTATUS_IBIWON) |
+			 BIT(NPCX_I3C_MSTATUS_NOWCNTLR));
```
Done

alvsun commented:
> With #70773 now merged... you'll need to add the compat macro to
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L88
> 
> and
> https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/drivers/i3c/i3c_shell.c#L99
> 
> in a separate commit

Done

alvsun commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2055 @@
+	if (config->common.dev_list.num_i3c == 0) {
+		LOG_ERR("%s: No i3c target device in the device tree", __func__);
+		return -ENODEV;
+	}
```
Done

fabiobaltieri commented on dts/arm/nuvoton/npcx/npcx4.dtsi:
```diff
@@ -14,6 +14,8 @@
 #include "npcx4/npcx4-espi-vws-map.dtsi"
 /* npcx4 series low-voltage io controls mapping table */
 #include "npcx4/npcx4-lvol-ctrl-map.dtsi"
+/* npcx4 series reset mapping table*/
```
`/* npcx4 series reset mapping table */`

(missing space before the comment closing marker)

alvsun replied:
Done

fabiobaltieri commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2053 @@
+	DEVICE_DT_INST_DEFINE(id, npcx_i3c_init, NULL, &npcx_i3c_data_##id, &npcx_i3c_config_##id, \
+			      POST_KERNEL, CONFIG_I3C_CONTROLLER_INIT_PRIORITY,                    \
+			      &npcx_i3c_driver_api);                                               \
+	static void npcx_i3c_config_func_##id(const struct device *dev)                            \
+	{                                                                                          \
+		IRQ_CONNECT(DT_INST_IRQN(id), DT_INST_IRQ(id, priority), npcx_i3c_isr,             \
+			    DEVICE_DT_INST_GET(id), 0);                                            \
+		irq_enable(DT_INST_IRQN(id));                                                      \
+	};
```
you could move this to the top of the macro, then you won't have to forward declare it

alvsun replied:
Done

XenuIsWatching commented on drivers/i3c/i3c_shell.c:
```diff
@@ -86,6 +86,7 @@ struct i3c_ctrl {
 
 DT_FOREACH_STATUS_OKAY(cdns_i3c, I3C_CTRL_FN)
 DT_FOREACH_STATUS_OKAY(nxp_mcux_i3c, I3C_CTRL_FN)
+DT_FOREACH_STATUS_OKAY(nuvoton_npcx_i3c, I3C_CTRL_FN)
```
nit: This should be in alphabetical order

alvsun replied:
Done

alvsun committed:
drivers: i3c: i3c_shell: add alphabetical order constraint

Add /* zephyr-keep-sorted-start */ around DT_FOREACH_STATUS_OKAY.

Signed-off-by: Alvis Sun <yfsun@nuvoton.com>

MulinChao reviewed (approved):
LGTM.

XenuIsWatching commented on drivers/i3c/i3c_npcx.c:
```diff
@@ -0,0 +1,2052 @@
+		goto out_do_ccc;
+	}
+
+	/* Write data (defining byte or data bytes) for CCC if needed */
+	if (payload->ccc.data_len > 0) {
+		npcx_i3c_status_clear_all(inst);
+		npcx_i3c_errwarn_clear_all(inst);
+		xfered_len = npcx_i3c_xfer_write_fifo(inst, payload->ccc.data,
+						      payload->ccc.data_len, false);
+		if (xfered_len < 0) {
+			LOG_ERR("CCC[0x%02x] %s command payload error (%d)", payload->ccc.id,
+				i3c_ccc_is_payload_broadcast(payload) ? "broadcast" : "direct",
+				ret);
+			ret = xfered_len;
+
+			goto out_do_ccc;
+		}
+
+		/* Write back the transferred bytes */
+		payload->ccc.num_xfer = xfered_len;
+	}
```
How does this distinguish the difference between the defining byte and payload for the direct CCC in I3C v1.1?

alvsun replied:
Only the i3c_ccc APIs will know the difference (e.g., i3c_ccc_do_getstatus()). 
This part handles both broadcast and direct CCC write data operations in the earlier stage(marked as red in the picture below).

![image](https://github.com/zephyrproject-rtos/zephyr/assets/150219805/7fd21301-98d6-4f0f-a63b-8f28fe121bbf)
